clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -417,7 +497,10 @@ public void unmute(String id) {
     }
 
     private void unmute(KafkaChannel channel) {
-        channel.unmute();
+        explicitlyMutedChannels.remove(channel);
+        if (!outOfMemory) { //only if we're not muting everything because no memory
+            channel.unmute();
```

Comment:
If this code was only called from tests, then channels would remain in `explicitlyMutedChannels` forever :-) It is actually called by the broker - mute/unmute to control reading from the channel and hence the need to track explicitly muted channels.